staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -437,35 +448,86 @@ func (p *patcher) applyPatch(_ context.Context, _, currentObject runtime.Object)
+		return nil, errors.NewBadRequest(fmt.Sprintf("the API version in the data (%s) does not match the expected API version (%v)", gvk.GroupVersion().String(), p.kind.GroupVersion().String()))
+	}
+	if err := checkName(objToCreate, p.name, p.namespace, p.namer); err != nil {
+		return nil, err
+	}
+	if _, err := p.applyAdmission(ctx, objToCreate, nil); err != nil {
+		return nil, err
+	}
+	attributes := p.admissionAttributes(ctx, objToCreate, nil)
+	createValidation := rest.AdmissionToValidateObjectFunc(p.admissionCheck, attributes)
+	return p.restApplier.Create(ctx, objToCreate, createValidation, false)
+}
+
+func (p *patchCreater) admissionType() admission.Operation {
+	return admission.Create
+}
+
+func (p *patcher) admissionAttributes(ctx context.Context, updatedObject runtime.Object, currentObject runtime.Object) admission.Attributes {
+	admissionOperation := p.operation.admissionType()
+	userInfo, _ := request.UserFrom(ctx)
+	return admission.NewAttributesRecord(updatedObject, currentObject, p.kind, p.namespace, p.name, p.resource, p.subresource, admissionOperation, userInfo)
+}
+
 // applyAdmission is called every time GuaranteedUpdate asks for the updated object,
 // and is given the currently persisted object and the patched object as input.
 func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Object, currentObject runtime.Object) (runtime.Object, error) {
 	p.trace.Step("About to check admission control")
-	return patchedObject, p.admissionCheck(patchedObject, currentObject)
+	mutatingAdmission, ok := p.admissionCheck.(admission.MutationInterface)
+	if ok && mutatingAdmission.Handles(p.operation.admissionType()) {
+		attributes := p.admissionAttributes(ctx, patchedObject, currentObject)
+		return patchedObject, mutatingAdmission.Admit(attributes)
+	}
+	return patchedObject, nil
 }
 
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, error) {
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
-		p.mechanism = &jsonPatcher{patcher: p}
+		mechanism := &jsonPatcher{patcher: p}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, err
 		}
-		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		mechanism := &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, model: scope.OpenAPISchema}
+		_, err := p.restPatcher.Get(ctx, p.name, &metav1.GetOptions{})
```

Comment:
this is racy… determining whether the create or update based on initial get means our determination can be wrong if a conflict/already exists/notfound is encountered and we need to retry

Reply:
a good test of this would be to simultaneously apply an object with a distinct annotation in X (maybe 100?) goroutines and delete it in a tight loop in another goroutine (ignoring notfound errors). Every apply should succeed either with a return status that indicated it was created or simply succeeded, and the result of each apply call should have the submitted annotation present.

Reply:
A patch request being a valid create request and a patch request being turned into a valid update are different things. Create is not guaranteed to always return "object already exists" if the object already exists, only if the object already exists and the create request was made with a valid object, all the validation in create is done before actually checking if the object exists or not. Going into the update path only when create returns "object already exists" could still be racy, I think, and going into it when create fails for any reason would make it so error messages from create (for things like a field being missing or having an invalid value) can never get back to the user.

I think the only way to really get around the race would be to:
1. Call get, if object already exists skip to step 4
2. Call create, return created object if successful
3. Call get, return error from the previous create if the get returns an "object not found" error
4. Call update

Reply:
calling Get outside `applyPatch` is what is racy. `applyPatch` gets re-called inside GuaranteedUpdate multiple times if the contents of etcd change.

today, `applyPatch` requires the object to exist:
```
	if hasUID, err := hasUID(currentObject); err != nil {
		return nil, err
	} else if !hasUID {
		return nil, errors.NewNotFound(p.resource.GroupResource(), p.name)
	}
```

if we were going to to relax that requirement, and handle doing create permission checks, etc, I think that would be where to do it